Dashboards Contracts sidebar filters

eltercero opened the pull request:
Closes #3047

## :v: What does this PR do?

It adds the sidebar filters to Dashboards Contracts.

@Crashillo I think this was the first time that Dropdown and Blockheader are being used together, and there were a couple of bugs. The first is a matter of styling, I added a new rule `widht: 100%` in `app/assets/stylesheets/comp-block-header.scss`. Otherwise the "All" link and the blockheader title would stay together (even with the flex and justify-between rules). The other bug was that when clicking "All" would also work as a toggle and collapse. The fix, which @jorgeatgu figured out, was adding the event modifier `stop` (as in `@click.stop="selectAll"`). Now it works.

What's still pending to do is using the Caret component in Blockheader and make it rotate accordingly.

### Some implementation details

I'd like to share some details so you can better understand the reasons behind some parts of the code. 

The most important thing you need to know is that all the filtering in the sidebars is actually made by the dc charts, so there's no filtering logic in the vue components.

To make this work, the dc charts have to be initialized no matter the page you are or start at. So, for instance, if you enter `/dashboards/contratos/contratos` the dc charts (which are not in that component) have to be created because you could filter from the sidebar there. 

Another issue that arised is that, since they could be created without the component being shown, when going to Summary the charts didn't have the proper size (the size is calculated when creating the chart using the size of the target element). So there's an event called `moved_to_summary` that will be listened just once to redraw the charts when going to Summary from others . 

Last, there's a sidebar filter for dates. To make this work, there's a hidden dc chart for dates that does all the filtering with the other charts.

## :mag: How should this be manually tested?

Head to https://getafe.gobify.net/dashboards/contratos/resumen and play with the sidebar filters. Case scenarios:

- Clicking on any sidebar filter will update the current filters. It should also update the chart bars if the filter has a char version.
- When filtering with the chart bar, the sidebar filter should update if it applies.
- If there's a version of the filter as a chart, the chart filter should be updated too.
- If you're using the sidebar filter from any index (contracts or tenders), listing should be updated

## :eyes: Screenshots

![Captura de pantalla 2020-05-21 a las 10 51 40](https://user-images.githubusercontent.com/545235/82541646-18443180-9b51-11ea-870f-d5e917b68b54.png)

## :shipit: Does this PR changes any configuration file?

No

(Changes in these files might need to update the role in Ansible)

No

eltercero committed:
Initial version of sidebars filters

- Date filter (without last 12 months for now, that's still wip), process type and contract type
- Vue app changes to allow pages to refresh from wherever the sidebar is used.
- Vue app changes to allow Summary component to exist at all time, since it's being used for filtering from every component.

eltercero committed:
Dropdown for the sidebar filters

eltercero committed:
Removing mean savings percentage

eltercero committed:
Fixing tests

eltercero commented:
@ferblape @Crashillo ready for review

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -6,6 +6,7 @@ import { scaleThreshold } from 'd3-scale';
 
 const d3 = { scaleThreshold, sum, mean, median, max }
 
+import * as dc from 'dc'
```
Is there any chance to code splitting `dc`?

eltercero replied:
It's actually not needed here, I forgot to remove it. I'll delete it.

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -137,23 +146,25 @@ export class ContractsController {
       contract.final_amount = final_amount;
       contract.initial_amount = initial_amount;
       contract.range = rangeFormat(+final_amount);
+      contract.start_date_year = contract.start_date != undefined && contract.start_date != '' ? (new Date(contract.start_date).getFullYear()) : contract.start_date;
```
To check for `undefined` and length, you can leave it as:

```suggestion
      contract.start_date_year = contract.start_date ? (new Date(contract.start_date).getFullYear()) : contract.start_date;
```
Undefineds, nulls, and empty lengths are evaluated to `false`

Ay, ay, ay, that destructuring...

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -258,10 +263,13 @@ export class ContractsController {
     }
 
-    new GroupPctDistributionBars(renderOptions);
+    charts['contract_types'] = new GroupPctDistributionBars(renderOptions);
```
I'm seeing you're declaring this file as a `Class`, then, why do you use *global* variables? Declara them in the constructor, and use them with `this`

eltercero replied:
Yeah, you're right. I'll do it in the latest branch to avoid conflicts, since this file has a lot of changes.

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -270,19 +278,58 @@ export class ContractsController {
+  }
+
+  _updateChartsFromFilter(options){
+    const container = charts[options.id].container;
+
+    if (options.all) {
+      container.filter(null);
+      container.filter([options.titles]);
+    } else {
+      container.filter(options.title);
     }
 
-    new GroupPctDistributionBars(renderOptions);
+    Object.values(charts).forEach((chart) => chart.container.redraw());
+  }
```
I hope `options` brings all these properties.... 😇

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -270,19 +278,58 @@ export class ContractsController {
-      status === 'Formalizado' || status === 'Adjudicado'
+  _formalizedContractsData(contractsData){
+    return contractsData.filter(({status}) =>
+      status == 'Formalizado' || status == 'Adjudicado'
```
What? What? What? Removing a best practice? OMG.... Goku, kill him!

eltercero replied:
Ahh, that's why there was a git clonflict there 😅

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,138 @@
+        :key="filter.id"
+        class="dashboards-home-aside--block"
+      >
+        <Dropdown @is-content-visible="filter.isToggle = !filter.isToggle">
```
Are we using isToggle somewhere?

Crashillo replied:
Maybe in `@toggle=toggle(filter)`? Plz, confirm

eltercero replied:
It was `@is-content-visible="toggle(filter)"`. Thanks!

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,138 @@
+          option.isOptionChecked = !option.isOptionChecked;
+        }
+      })
+    });
```
What if the `Aside` parent component is who listen the events and pass the props down? Asking

eltercero replied:
In the first case I think it makes sense, I changed it. I don't think it'd work well for the second case, though, it'll be more complex.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,138 @@
+
+        counter.dates[contract.start_date_year] = counter.dates[contract.start_date_year] || 0
+        counter.dates[contract.start_date_year]++
+      })
```
I've seen this before....

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Home.vue:
```diff
@@ -1,17 +1,19 @@
+          <ContractsIndex v-show="isCurrentPath('contracts_index')"/>
+          <ContractsShow v-if="isCurrentPath('contracts_show')"/>
+          <TendersIndex v-show="isCurrentPath('tenders_index')"/>
+          <TendersShow v-if="isCurrentPath('tenders_show')"/>
```
It isn't a good practice adding function calls in the DOM. Every single time the component is updated, therefore, rendered; is calling to these functions. That penalizes the performance.

You should declare data properties or computed functions. Due to the name you're using, I'm assuming `isCurrentPath` runs at mounted, and stills.
If so, declare boolean flags to handle this.

eltercero replied:
Changed to computed properties 👍

Crashillo commented on app/javascript/gobierto_dashboards/webapp/lib/config.js:
```diff
@@ -16,3 +16,35 @@ export const assigneesColumns = [
+        {id: 2, title: 'Basado en Acuerdo Marco', counter: 0, isOptionChecked: false},
+        {id: 3, title: 'Negociado con publicidad', counter: 0, isOptionChecked: false},
+        {id: 4, title: 'Negociado sin publicidad', counter: 0, isOptionChecked: false}
+      ]
```
Untraslated stuff? Oh lord

eltercero replied:
That's the title coming from the csv. I did it this way so make things simple, but maybe I should generate it automatically based on the csv :\

ferblape replied:
@eltercero yes, I think you should make these lists dynamic from the CSV, there might be more types

eltercero replied:
Changed!

eltercero committed:
Removing dc import from contracts controller

eltercero committed:
No need to check if empty string or undefined since both are false

Co-authored-by: Hugoren Martinako <aumpfbahn@gmail.com>

eltercero committed:
Merge branch '3047-sidebar-filters' of github.com:PopulateTools/gobierto into 3047-sidebar-filters

eltercero committed:
Using triple equal operator

eltercero committed:
contractsData is now passed as a prop to Aside from Home

And home is responsible for listening the events regarding contractsData.

eltercero committed:
Destructuring

eltercero committed:
Changed function to computed properties to improve performance

eltercero committed:
Dinamically generate process types and contract types options

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -91,6 +91,14 @@ export class ContractsController {
+
+        EventBus.$once('moved_to_summary', () => {
+          this._redrawCharts();
+        });
```
Try to include an explanatory comment for each event. Specially the events. You may do it in following PRs

Crashillo replied:
🤔 Indeed, why the last event is `$once` for? I mean, I know what `once` does, but what purpose?

eltercero replied:
I'll add an explanation for that.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,169 @@
+        if (option.title === title) {
+          option.isOptionChecked = !option.isOptionChecked;
+        }
+      })
```
High error possibility

eltercero replied:
What do you suggest?

Crashillo replied:
```suggestion
      const { options = [] } = this.filters.find(( { id: i } ) => id === i) || {};

      options.forEach(option => {
        if (option.title === title) {
          option.isOptionChecked = !option.isOptionChecked;
        }
      })
```

eltercero replied:
👏 thanks, it's much better and it works great.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,169 @@
+  },
+  methods: {
+    initFilterOptions(){
+      const contractTypesOptions = [], processTypesOptions = [], dateOptions = [];
```
Different lines

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,169 @@
+      })
+    },
+    updateCounters(firstUpdate=false) {
+      const counter = {process_types: {}, contract_types: {}, dates: {}};
```
Different lines

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,169 @@
+    updateCounters(firstUpdate=false) {
+      const counter = {process_types: {}, contract_types: {}, dates: {}};
+
+      debugger
```
Hello?

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,169 @@
+
+      debugger
+
+      this.contractsData.forEach(({process_type, contract_type, start_date_year}) => {
+        counter.process_types[process_type] = counter.process_types[process_type] || 0
+        counter.process_types[process_type]++
+
+        counter.contract_types[contract_type] = counter.contract_types[contract_type] || 0
+        counter.contract_types[contract_type]++
+
+        counter.dates[start_date_year] = counter.dates[start_date_year] || 0
+        counter.dates[start_date_year]++
+      })
```
Explain this function

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Aside.vue:
```diff
@@ -1,14 +1,169 @@
+    },
+    toggle(filter){
+      this.filters.forEach(_filter => {
+        if (_filter.id == filter.id) {
```
===

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/shared/Home.vue:
```diff
@@ -21,25 +25,56 @@
+    isContractsIndex: function() { return this.$route.name === 'contracts_index' },
+    isContractsShow: function() { return this.$route.name === 'contracts_show' },
+    isTendersIndex: function() { return this.$route.name === 'tenders_index' },
+    isTendersShow: function() { return this.$route.name === 'tenders_show' },
```
1. ES6
2. Are these values changing while the component is mounted? It looks like they don't

eltercero replied:
1. I think you can't use arrow functions in computed properties, or do you mean something else?

2. I think they are changing. Home is mounted just once and every time the tab changes these properties change too.

Crashillo replied:
I'm just saying so, keep consistency in the code format

```suggestion
    isSummary() { return this.$route.name === 'summary' },
    ...etc
```

eltercero replied:
Ahh. Understood, thanks!

eltercero committed:
Removing forgotten debugger, adding explanatory comment for a function

eltercero committed:
Triple equals

eltercero committed:
Some improvements

- Adding explanation to an event.
- Improving readability in variable declaration.

eltercero committed:
Improving function on dc filter selected event

Co-authored-by: Hugoren Martinako <aumpfbahn@gmail.com>

eltercero committed:
Refactor to maintain consistency
